python/ray/workflow/step_executor.py

```diff
@@ -194,17 +216,68 @@ async def _write_step_inputs(wf_storage: workflow_storage.WorkflowStorage,
     await asyncio.gather(*save_tasks)
 
 
+def _validate_checkpoint_options(ret: Union["Workflow", Any]):
+    # The condition where checkpointing is not allowed.
+    if not isinstance(ret, Workflow):
+        return
+    context = workflow_context.get_workflow_step_context()
+    if context is None:
+        checkpoint_context = workflow_context.CheckpointContext()
+    else:
+        checkpoint_context = context.checkpoint_context
+    detached_from_dag = checkpoint_context.detached_from_dag
+    if (not detached_from_dag
+            and checkpoint_context.checkpoint != CheckpointMode.SKIP.value):
+        return
+
+    invalid_step = None
+    for w in ret._iter_workflows_in_dag():
+        if w.data.step_options.checkpoint:
+            if w != ret or detached_from_dag:
```

Comment:
```suggestion
        if w.data.step_options.checkpoint:
            # If the outer workflow step is skipping checkpointing, we need to make sure that any inner workflow steps are not detached from the top-level DAG.
            if w != ret:
              # If this is not the output step of the sub-workflow, then it will be detached from the main DAG.
            if detached_from_dag:
              # The outer workflow is already detached, so we definitely can't checkpoint any of its sub-workflow steps.
```